reference types for typed functions

WhaleKit opened the pull request:
Currently neither `Extern::typed_func` nor `ModuleInstance::exported_func` can have references from "reference types" proposal in their signatures. In this PR i'm adding 2 newtypes for that usage, with from/into underlying u32, but not deref.

I briefly considered making `FuncAddr` and `ExternAddr` themselves into newtypes (from aliases), but that would break existing public api.

Also includes example with function reference being passed around as a callback

explodingcamera commented:
Thanks for the PR! I'll have to think about this a bit more and check how other engines handle this. I think replacing the FuncAddr and ExternAddr in WasmValue with a newtype (e.g FuncRef and ExternRef) is probably the way to go, as the address should be opaque (the user probably shouldn't be able to pass in their own u32), and the next version will have breaking changes anyways.

WhaleKit commented:
Updated. This PR now only has example of using function references in typed (exported and imported) functions, which I updated to reflect the new types.
